server/users.ts

```diff
@@ -1040,12 +1040,19 @@ export class User extends Chat.MessageContext {
 			Rooms.get(roomid)!.onLeave(oldUser);
 		}
 
-		if (this.locked === '#dnsbl' && !oldUser.locked) this.locked = null;
-		if (!this.locked && oldUser.locked === '#dnsbl') oldUser.locked = null;
+		const startLock = this.locked;
+		const startSemilock = this.semilocked;
+
+		if (this.semilocked === '#dnsbl' && !oldUser.semilocked) this.semilocked = null;
```

Comment:
I wouldn't propagate any semilock, actually.

```suggestion
		if (!oldUser.semilocked) this.semilocked = null;
```